File: benchmarks/collection-suite/benchmark.py
import sys


def vector_push_sum():
    N = 1_000_000
    lst = []
    for i in range(N):
        lst.append(i * 17 + 3)
    checksum = sum(lst)
    for _ in range(500_000):
        lst.pop()
    checksum += len(lst)
    return checksum


def vector_sort():
    SIZE = 100_000
    mask = 0xFFFFFFFF
    state = 123456789
    values = []
    for _ in range(SIZE):
        state = (state * 1664525 + 1013904223) & mask
        # sign-extend to int32
        v = state if state < 0x80000000 else state - 0x100000000
        values.append(v)
    # insertion sort
    for i in range(1, SIZE):
        key = values[i]
        j = i - 1
        while j >= 0 and values[j] > key:
            values[j + 1] = values[j]
            j -= 1
        values[j + 1] = key
    checksum = sum(values)
    checksum += values[0] * 31
    checksum += values[SIZE - 1] * 17
    return checksum


def hashmap_insert_lookup():
    N = 200_000
    d = {}
    for i in range(N):
        k = i * 17 + 31
        v = i * 13 + 7
        d[k] = v
    checksum = 0
    for i in range(N):
        k = i * 17 + 31
        if k in d:
            checksum += d[k]
    return checksum


BENCHMARKS = {
    'vector-push-sum': vector_push_sum,
    'vector-sort': vector_sort,
    'hashmap-insert-lookup': hashmap_insert_lookup,
}

if len(sys.argv) != 2 or sys.argv[1] not in BENCHMARKS:
    raise SystemExit(2)
print(BENCHMARKS[sys.argv[1]]())

File: benchmarks/collection-suite/benchmark.js
'use strict';
const arg = process.argv[2];

function vectorPushSum() {
    const N = 1_000_000;
    const list = [];
    for (let i = 0; i < N; i++) list.push(i * 17 + 3);
    let checksum = BigInt(0);
    for (let i = 0; i < N; i++) checksum += BigInt(list[i]);
    for (let i = 0; i < 500_000; i++) list.pop();
    checksum += BigInt(list.length);
    return checksum;
}

function vectorSort() {
    const SIZE = 100_000;
    const values = new Int32Array(SIZE);
    let state = 123456789 >>> 0;
    for (let i = 0; i < SIZE; i++) {
        state = Math.imul(state, 1664525) + 1013904223 >>> 0;
        values[i] = state | 0;
    }
    // insertion sort
    for (let i = 1; i < SIZE; i++) {
        const key = values[i];
        let j = i - 1;
        while (j >= 0 && values[j] > key) { values[j + 1] = values[j]; j--; }
        values[j + 1] = key;
    }
    let checksum = BigInt(0);
    for (let i = 0; i < SIZE; i++) checksum += BigInt(values[i]);
    checksum += BigInt(values[0]) * BigInt(31);
    checksum += BigInt(values[SIZE - 1]) * BigInt(17);
    return checksum;
}

function hashmapInsertLookup() {
    const N = 200_000;
    const map = new Map();
    for (let i = 0; i < N; i++) {
        const k = i * 17 + 31;
        const v = i * 13 + 7;
        map.set(k, v);
    }
    let checksum = BigInt(0);
    for (let i = 0; i < N; i++) {
        const k = i * 17 + 31;
        const v = map.get(k);
        if (v !== undefined) checksum += BigInt(v);
    }
    return checksum;
}

let result;
if (arg === 'vector-push-sum') result = vectorPushSum();
else if (arg === 'vector-sort') result = vectorSort();
else if (arg === 'hashmap-insert-lookup') result = hashmapInsertLookup();
else process.exit(2);

console.log(result.toString());
